### PR TITLE
Add SSH failure callbacks

### PR DESCRIPTION
**Pull request by imcdo**
One common point of failure for ducktape is paramiko, when a remote account fails, we generally get a non useful error message that has no attachment to the node that is being ran.  This PR addresses this issue by adding a flag to ducktape that allows you to specify a specific function that takes in an error and a remote account, that will then be ran on an ssh failure.  For instance, if you are running aws instances, you can write your own aws command that validates that the node is still running on aws, and then run this on an ssh failure.

### CODE REVIEW

File: ducktape/cluster/json.py
Comment (sonupillai) [inline ducktape/cluster/json.py:104]: This is a big change. Please make sure, all callers of this function are updated to access "externally_routable_ip"  config like a kw?

**Review by sonupillai — APPROVED** [review]
LGTM and as we discussed but I would recommend getting a second review from Ducktape Gods as well.

File: ducktape/cluster/json.py
Comment (imcdo) [reply ducktape/cluster/json.py:104]: they dont need to, as args also get forwarded.

File: docs/debug_tests.rst
Comment (stan-confluent) [inline docs/debug_tests.rst:101]: nit: SSH should be all caps, no?

File: ducktape/cluster/remoteaccount.py
Comment (stan-confluent) [inline ducktape/cluster/remoteaccount.py:43]: should it be `raise e`? I don't remember if simple `raise` would preserve the original traceback

File: ducktape/command_line/main.py
Comment (stan-confluent) [inline ducktape/command_line/main.py:187]: There needs to be a readable error message if the function cannot be loaded for some reason.

File: ducktape/cluster/linux_remoteaccount.py
Comment (stan-confluent) [inline ducktape/cluster/linux_remoteaccount.py:22]: there used to be a `logger` arg, but I don't see it being passed in in the `make_remote_account` method. Was anyone using it?

File: tests/cluster/check_remoteaccount.py
Comment (stan-confluent) [inline tests/cluster/check_remoteaccount.py:96]: I'm presuming other existing tests provide coverage for when there are no exception checks? Please make sure that we have covered the cases when there's more than one exception check, no exception checks or invalid exception check(s).

File: ducktape/command_line/parse_args.py
Comment (stan-confluent) [inline ducktape/command_line/parse_args.py:80]: Nit: Can this flag be specified multiple times or do you specify multiple functions after a single flag? It's not clear from the description

**Review by stan-confluent — COMMENTED** [review]
Thanks Ian, this is great - great idea, clean implementation, 👏 
Couple of comments from me, mostly around testing and negative scenarios, and the usual batch of nitpicks, but looks good overall.

File: ducktape/cluster/remoteaccount.py
Comment (imcdo) [reply ducktape/cluster/remoteaccount.py:43]: i originally did `raise e` however when getting the traceback from pytest, it would show me the call stack from inside here, while with just raise, it gives the original stack trace, though i havn't tried this in python3

File: tests/cluster/check_remoteaccount.py
Comment (imcdo) [reply tests/cluster/check_remoteaccount.py:96]: No exception checks are the default case, so they are covered, will add the multiple case specificly

File: ducktape/command_line/main.py
Comment (imcdo) [reply ducktape/command_line/main.py:187]: added to the load function method

File: ducktape/cluster/remoteaccount.py
Comment (imcdo) [reply ducktape/cluster/remoteaccount.py:43]: https://stackoverflow.com/questions/36153805/difference-between-raise-and-raise-e

File: ducktape/cluster/linux_remoteaccount.py
Comment (stan-confluent) [reply ducktape/cluster/linux_remoteaccount.py:22]: Nah, I was mostly making sure that you don't get any unintentional consequences of passing all the args from the method above, but I don't see a real issue anywhere

File: ducktape/cluster/remoteaccount.py
Comment (stan-confluent) [reply ducktape/cluster/remoteaccount.py:43]: As long as it works fine in py3 (which is what newer ducktape versions are all targeting anyway), I'm fine with it.

File: ducktape/utils/util.py
Comment (stan-confluent) [inline ducktape/utils/util.py:63]:
```python
except AttributeError as e:
    raise Exception(...) from e
```

Comment (imcdo) [reply ducktape/utils/util.py:63]: doesn't look like this is compatable with python2 :/ 
```
try:
...     raise Exception('hi')
... except Exception as e:
...     raise Exception('there') from e
  File "<stdin>", line 4
    raise Exception('there') from e
                                ^
SyntaxError: invalid syntax

```

Comment (stan-confluent) [reply ducktape/utils/util.py:63]: _mumbles swear words into the beard_

Comment (imcdo) [reply ducktape/utils/util.py:63]: python2 man >:(

File: ducktape/command_line/parse_args.py
Comment (stan-confluent) [reply ducktape/command_line/parse_args.py:80]: cool, as long as it is clear from the docs. I was confused because actual checkers on the cluster object are a list

**Review by stan-confluent — APPROVED** [review]

File: ducktape/cluster/remoteaccount.py
Comment (ewencp) [inline ducktape/cluster/remoteaccount.py:42]: This should also be wrapped in a try/except if we support multiple checks.

**Review by ewencp — COMMENTED** [review]
Does this actually need to be configurable, or is it something that can be baked in? Even if just providing a few options ootb?

The description is about the error logging, but that could be improved without a configurable validator. What type of validation would be happening other than trying to reconnect via the remote account object or just logging more detailed output? The main thing I can think of would be checks against e.g. a cloud provider, but since that info is injected after allocation atm, a check like that would need more substantial changes.

File: ducktape/cluster/remoteaccount.py
Comment (imcdo) [reply ducktape/cluster/remoteaccount.py:42]: The current assumption is that the validators will raise an exception if something is wrong.  Should this paradigm change?  Should it return some sort of result object?  We can also just try catch it but we can only raise one error in python2 (and python3 its a chain of exceptions which is a different assumption AFAIK)

**Comment by imcdo** [comment]
> Does this actually need to be configurable, or is it something that can be baked in? Even if just providing a few options ootb?
I think this is better to be left as something that is configurable, as  ducktape isn't responsible for the environment at all and operates in the dark outside of ssh access.  Relying on the user to pass in validation about the environment when ssh fails as they are responsible for creating the environment allows for validation a better understanding of the environment.

File: ducktape/cluster/remoteaccount.py
Comment (ewencp) [reply ducktape/cluster/remoteaccount.py:42]: Yeah, you can only raise one exception, but you can make an exception type that wraps others. I just wasn't sure what types of checks you might do and in what order. Since this approach requires doing only the most specific checks, it can be more expensive (e.g. checking if a cloud instance is destroyed vs just doing a quick `nc -z` check first).

I'm not sure how many use cases there would be for using multiple checks other than a) fast -> slow ordering for checks and b) if you have multiple detectable failure modes. For (b), I'm not sure what that list would be.

**Review by ewencp — APPROVED** [review]
ok, guess it makes sense for the shortest path. the reason i ask is that one alternative would be a bigger change, but require less configuration -- tie the handler to the Cluster/ClusterNode and just have a native implementations for things like cloud providers. In particular, vagrant has been painful for long running clusters for a long time and we effectively bake certain implementations into ducktape anyway (in particular vagrant).

I don't mind making things pluggable, but where possible it'd be nice to keep the settings to a minimum if they're effectively going to be identical for a given cluster type. Another way to think of this in the current state is that if an instance disappears, vagrant status should be able to tell you that too, so even without adding custom cloud providers you could still bake the checks into the Cluster implementation.

More broadly, I was wasn't sure how this would work generally given the limited info that is passed along via the ClusterNode (other than trying some simpler tests like nc -z but that may fail just because of network connectivity, not only if the cloud resource was destroyed). It means you have to be able to backtrack to the original information about the node. That's possible in some cases, I'm not sure about all of them, and even if it is, could be relatively expensive.

We can make this change, but I do think we should continue to consider a more integrated and native approach for different cluster types. e.g. we had thought about a native aws implementation, but vagrant was the fastest route originally to support both local + aws.

**Pull request merged**
